chainer/functions/math/sparse_matmul.py

```diff
@@ -0,0 +1,425 @@
+            ret.append(g_dn)
+        return ret
+
+
+def _sparse_matmul_gradsp(a, b, c_row, c_col, c_shape,
+                          transa, transb, transc, dtype):
+    if dtype is None:
+        dtype = numpy.result_type(a.dtype, b.dtype)
+
+    if transa:
+        A = a.swapaxes(-1, -2)
+    else:
+        A = a
+    if transb:
+        B = b.swapaxes(-1, -2)
+    else:
+        B = b
+    if transc:
+        C_row = c_col
+        C_col = c_row
+    else:
+        C_row = c_row
+        C_col = c_col
+
+    xp = cuda.get_array_module(A, B)
+    if xp is numpy:
+        return _sparse_matmul_gradsp_cpu(A, B, C_row, C_col, dtype)
+    else:
+        return _sparse_matmul_gradsp_gpu(A, B, C_row, C_col, dtype)
+
+
+def _sparse_matmul_gradsp_cpu(A, B, C_row, C_col, dtype):
+    # A.shape: ((nb,) _m, _k)
+    # B.shape: ((nb,) _k, _n)
+    # C_row/col.shape: ((nb,) ldnz)
+    _m, _k = A.shape[-2:]
+    ldnz = C_row.shape[-1]
+    C = numpy.matmul(A, B).astype(dtype, copy=False)
+    if A.ndim == 2:
+        C_data = numpy.zeros((ldnz), dtype=dtype)
+        nnz = len(numpy.where(C_row >= 0)[0])
+        C_data[:nnz] = C[C_row[:nnz], C_col[:nnz]]
+    else:
+        nb = A.shape[0]
+        C_data = numpy.zeros((nb, ldnz), dtype=dtype)
+        for i in range(nb):
+            nnz = len(numpy.where(C_row[i] >= 0)[0])
+            C_data[i, :nnz] = C[i, C_row[i, :nnz], C_col[i, :nnz]]
+
+    return C_data
+
+
+def _sparse_matmul_gradsp_gpu(A, B, C_row, C_col, dtype):
+    # A.shape: ((nb,) _m, _k)
+    # B.shape: ((nb,) _k, _n)
+    # C_row/col.shape: ((nb,) ldnz)
+    _m, _k = A.shape[-2:]
+    _n = B.shape[-1]
+    ldnz = C_row.shape[-1]
+    if A.ndim == 2:
+        nb = 1
+        C_data = cuda.cupy.zeros((ldnz), dtype=dtype)
+    else:
+        nb = A.shape[0]
+        C_data = cuda.cupy.zeros((nb, ldnz), dtype=dtype)
+
+    nthreads = nb * ldnz
+    _cupy_sparse_matmul_gradsp()(nb, _m, _n, _k, ldnz, A, B,
+                                 C_data, C_row, C_col, size=nthreads)
+
+    return C_data
+
+
+def _cupy_sparse_matmul_gradsp():
+    return cuda.cupy.ElementwiseKernel(
+        'int32 nb, int32 _m, int32 _n, int32 _k, int32 nnz, \
+         raw A _A, raw B _B, \
+         raw C C_data, raw T C_row, raw T C_col',
+        '',
+        '''
+        int i_nz = (i % nnz);
+        int i_b = (i / nnz);
+        if (i_b >= nb) {
+            continue;
+        }
+        int i_C = i;
+        int i_m = C_row[i_C];
+        if (i_m < 0) {
+            continue;
+        }
+        assert(i_m < _m);
+        int i_n = C_col[i_C];
+        if (i_n < 0) {
+            continue;
+        }
+        assert(i_n < _n);
+        C val_C = 0.0;
+        for (int i_k = 0; i_k < _k; i_k++) {
+            int i_A = i_k + _k * (i_m + _m * i_b);
+            int i_B = i_n + _n * (i_k + _k * i_b);
+            A val_A = _A[i_A];
+            B val_B = _B[i_B];
+            val_C += (C)(val_A * val_B);
```

Comment:
I don't think it's an issue here but how about preferring `static_cast` in the elementwise kernels for safety?

Reply:
OK, will use `static_cast`.